# Chương 3 - Giới thiệu Keras và TensorFlow

## Mục tiêu học tập

- Hiểu TensorFlow là gì.
- Hiểu Keras là gì và mối quan hệ giữa Keras với TensorFlow.
- Biết các lựa chọn thiết lập môi trường học Deep Learning.
- Làm quen với tensor, variable, tensor operations và GradientTape.
- Thực hành xây dựng một mô hình Keras đơn giản.

## 3.1 TensorFlow là gì?

TensorFlow là nền tảng machine learning mã nguồn mở, hỗ trợ thao tác trên tensor, tự động tính gradient, chạy trên CPU/GPU/TPU và hỗ trợ triển khai mô hình.

Trong chương này, ta xem TensorFlow như lớp hạ tầng toán học bên dưới của Deep Learning.

In [3]:
import tensorflow as tf

print('TensorFlow version:', tf.__version__)
print('Số GPU khả dụng:', len(tf.config.list_physical_devices('GPU')))
print('Danh sách thiết bị GPU:', tf.config.list_physical_devices('GPU'))

TensorFlow version: 2.20.0
Số GPU khả dụng: 0
Danh sách thiết bị GPU: []


## 3.2 Keras là gì?

Keras là API cấp cao dùng để xây dựng mô hình Deep Learning nhanh và dễ hơn.

TensorFlow cung cấp hạ tầng tính toán.

Keras cung cấp giao diện thân thiện để tạo layer, model, optimizer, loss và training loop ở mức cao.

## 3.3 Keras và TensorFlow: Lược sử ngắn

Keras được phát hành vào tháng 03/2015, sớm hơn TensorFlow khoảng 8 tháng.

Ban đầu Keras chạy trên Theano. Sau đó Keras hỗ trợ nhiều backend như Theano, TensorFlow, CNTK, MXNet.

Từ năm 2018, Keras trở thành high-level API chính thức của TensorFlow.

Ngày nay, khi nói đến Keras trong thực hành, ta thường dùng `tensorflow.keras`.

## 3.4 Thiết lập môi trường Deep Learning

Các lựa chọn phổ biến để học và thực hành Deep Learning gồm:

1. Máy cá nhân có GPU NVIDIA.
2. Cloud như Google Cloud hoặc AWS.
3. Google Colab.

Google Colab phù hợp cho người mới vì không cần cài đặt và có thể dùng GPU miễn phí ở mức cơ bản.

Khuyến nghị: nếu chỉ học và demo, Colab là lựa chọn rất tốt. Nếu nghiên cứu nhiều và lâu dài, nên cân nhắc workstation hoặc cloud.

## 3.5.1 Tensor hằng và Variable

TensorFlow có tensor hằng và tensor có thể thay đổi trạng thái (`tf.Variable`).

Trong Deep Learning, trọng số của mô hình thường được lưu trong `Variable`.

In [4]:
x = tf.ones(shape=(2, 2))
y = tf.zeros(shape=(2, 2))
z = tf.random.normal(shape=(2, 2), mean=0.0, stddev=1.0)

print('Tensor x =')
print(x)
print('\nTensor y =')
print(y)
print('\nTensor z =')
print(z)
# Tạo một tensor 2x2 với các giá trị ngẫu nhiên phân bố đều 
# trong khoảng mặc định [0, 1).
v = tf.Variable(initial_value=tf.random.uniform(shape=(2, 2)))
print('\nVariable ban đầu =')
print(v)

v.assign(tf.ones((2, 2)))
print('\nVariable sau assign =')
print(v)

v.assign_add(tf.ones((2, 2)))
print('\nVariable sau assign_add =')
print(v)

Tensor x =
tf.Tensor(
[[1. 1.]
 [1. 1.]], shape=(2, 2), dtype=float32)

Tensor y =
tf.Tensor(
[[0. 0.]
 [0. 0.]], shape=(2, 2), dtype=float32)

Tensor z =
tf.Tensor(
[[ 0.6202532  -0.47684315]
 [ 0.67761695  1.7721741 ]], shape=(2, 2), dtype=float32)

Variable ban đầu =
<tf.Variable 'Variable:0' shape=(2, 2) dtype=float32, numpy=
array([[0.06050158, 0.7337326 ],
       [0.46755242, 0.03774917]], dtype=float32)>

Variable sau assign =
<tf.Variable 'Variable:0' shape=(2, 2) dtype=float32, numpy=
array([[1., 1.],
       [1., 1.]], dtype=float32)>

Variable sau assign_add =
<tf.Variable 'Variable:0' shape=(2, 2) dtype=float32, numpy=
array([[2., 2.],
       [2., 2.]], dtype=float32)>


## 3.5.2 Tensor operations

TensorFlow hỗ trợ nhiều phép toán quen thuộc như cộng, nhân ma trận, bình phương, căn bậc hai, giá trị lớn nhất theo từng phần tử.

Đây là nền tảng để xây dựng các phép biến đổi trong mạng nơ-ron.

In [5]:
# Tạo hai tensor hằng (constant) dạng ma trận 2x2
a = tf.constant([[1.0, 2.0], [3.0, 4.0]])
b = tf.constant([[5.0, 6.0], [7.0, 8.0]])

# Phép cộng hai tensor (cộng từng phần tử)
print('a + b =')
print(a + b)

# Bình phương từng phần tử của tensor a
print('\ntf.square(a) =')
print(tf.square(a))

# Nhân ma trận (matrix multiplication) giữa a và b
print('\ntf.matmul(a, b) =')
print(tf.matmul(a, b))

# Lấy giá trị lớn hơn giữa a và b theo từng phần tử
print('\ntf.maximum(a, b) =')
print(tf.maximum(a, b))

a + b =
tf.Tensor(
[[ 6.  8.]
 [10. 12.]], shape=(2, 2), dtype=float32)

tf.square(a) =
tf.Tensor(
[[ 1.  4.]
 [ 9. 16.]], shape=(2, 2), dtype=float32)

tf.matmul(a, b) =
tf.Tensor(
[[19. 22.]
 [43. 50.]], shape=(2, 2), dtype=float32)

tf.maximum(a, b) =
tf.Tensor(
[[5. 6.]
 [7. 8.]], shape=(2, 2), dtype=float32)


## 3.5.3 GradientTape và tự động vi phân

`GradientTape` cho phép TensorFlow theo dõi các phép tính để tự động tính gradient.

Đây là cơ chế quan trọng trong huấn luyện mạng nơ-ron bằng gradient descent.

In [6]:
# Khởi tạo biến w với giá trị ban đầu = 3.0 (có thể thay đổi trong quá trình tính toán)
w = tf.Variable(3.0)

# GradientTape dùng để ghi lại các phép toán nhằm phục vụ tính đạo hàm (gradient)
with tf.GradientTape() as tape:
    # Định nghĩa hàm y theo w (hàm bậc hai)
    y = w ** 2 + 2 * w + 1

# Tính đạo hàm của y theo w (dy/dw)
grad = tape.gradient(y, w)

# In ra giá trị của hàm y
print('Giá trị y =', y.numpy())

# In ra gradient (đạo hàm) tại w = 3
print('Gradient dy/dw =', grad.numpy())

Giá trị y = 16.0
Gradient dy/dw = 8.0


## Ví dụ trực quan về Gradient Descent

Ta tối ưu một hàm đơn giản: `loss = (w - 4)^2`.

Mục tiêu là cập nhật `w` để giá trị tiến dần về 4.

In [7]:
# Khởi tạo biến w với giá trị ban đầu = 0.0 (điểm xuất phát)
w = tf.Variable(0.0)

# Đặt learning rate (tốc độ học)
learning_rate = 0.1

# Lặp 10 bước để thực hiện gradient descent
for step in range(10):
    
    # Ghi lại các phép toán để tính gradient
    with tf.GradientTape() as tape:
        # Hàm loss cần tối ưu (mục tiêu là đưa w tiến về 4)
        loss = (w - 4.0) ** 2
    
    # Tính gradient của loss theo w (đạo hàm)
    grad = tape.gradient(loss, w)
    
    # Cập nhật w theo công thức gradient descent
    # w = w - learning_rate * gradient
    w.assign_sub(learning_rate * grad)
    
    # In thông tin mỗi bước: giá trị w, loss và gradient
    print(f'Bước {step:02d}: w = {w.numpy():.4f}, loss = {loss.numpy():.4f}, grad = {grad.numpy():.4f}')

Bước 00: w = 0.8000, loss = 16.0000, grad = -8.0000
Bước 01: w = 1.4400, loss = 10.2400, grad = -6.4000
Bước 02: w = 1.9520, loss = 6.5536, grad = -5.1200
Bước 03: w = 2.3616, loss = 4.1943, grad = -4.0960
Bước 04: w = 2.6893, loss = 2.6844, grad = -3.2768
Bước 05: w = 2.9514, loss = 1.7180, grad = -2.6214
Bước 06: w = 3.1611, loss = 1.0995, grad = -2.0972
Bước 07: w = 3.3289, loss = 0.7037, grad = -1.6777
Bước 08: w = 3.4631, loss = 0.4504, grad = -1.3422
Bước 09: w = 3.5705, loss = 0.2882, grad = -1.0737


## 3.6 Giải phẫu một mạng nơ-ron trong Keras

Các thành phần cốt lõi:

- `Layer`: khối xây dựng cơ bản.
- `Model`: tập hợp các layer.
- `compile()`: cấu hình optimizer, loss, metrics.
- `fit()`: huấn luyện mô hình.
- `evaluate()`: đánh giá.
- `predict()`: suy luận.

In [10]:
# Import Keras từ TensorFlow
from tensorflow import keras

# Import module layers để xây dựng các lớp (layer) trong mạng neural
from tensorflow.keras import layers

# Khởi tạo mô hình Sequential (các layer xếp chồng tuần tự)
model = keras.Sequential([
    
    # Layer 1: Dense (fully connected)
    # 16 neuron, hàm kích hoạt ReLU
    # input_shape=(8,) nghĩa là mỗi mẫu dữ liệu có 8 đặc trưng đầu vào
    layers.Dense(16, activation='relu', input_shape=(8,)),
    
    # Layer 2 (output layer):
    # 1 neuron, dùng sigmoid cho bài toán phân loại nhị phân
    layers.Dense(1, activation='sigmoid')
])

# Hiển thị cấu trúc mô hình (số layer, số tham số, output shape...)
model.summary()

Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_2 (Dense)                 │ (None, 16)             │           144 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_3 (Dense)                 │ (None, 1)              │            17 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 161 (644.00 B)

 Trainable params: 161 (644.00 B)

 Non-trainable params: 0 (0.00 B)

## Tạo dữ liệu giả để minh họa huấn luyện

Ở đây ta không dùng dữ liệu thật mà tạo dữ liệu nhị phân đơn giản để minh họa quy trình `compile()`, `fit()`, `evaluate()` và `predict()`.

In [11]:
# Import thư viện NumPy để xử lý dữ liệu
import numpy as np

# Đặt seed để kết quả random có thể tái lập (reproducible)
np.random.seed(42)

# Tạo dữ liệu train gồm 200 mẫu, mỗi mẫu có 8 đặc trưng
x_train = np.random.random((200, 8)).astype('float32')

# Tạo nhãn (label) cho train:
# Nếu tổng các đặc trưng > 4 thì label = 1, ngược lại = 0
y_train = (x_train.sum(axis=1) > 4).astype('float32')

# Tạo dữ liệu test gồm 50 mẫu, mỗi mẫu có 8 đặc trưng
x_test = np.random.random((50, 8)).astype('float32')

# Tạo nhãn cho test theo cùng logic như train
y_test = (x_test.sum(axis=1) > 4).astype('float32')

In [12]:
# Compile model: cấu hình cách học của mô hình
model.compile(
    
    # Optimizer: thuật toán tối ưu (cập nhật trọng số)
    # rmsprop là optimizer phổ biến cho Deep Learning
    optimizer='rmsprop',
    
    # Loss function: hàm mất mát
    # binary_crossentropy dùng cho bài toán phân loại nhị phân
    loss='binary_crossentropy',
    
    # Metrics: các chỉ số để đánh giá model trong quá trình train
    # accuracy = độ chính xác
    metrics=['accuracy']
)

# Huấn luyện mô hình
history = model.fit(
    
    # Dữ liệu đầu vào (features)
    x_train,
    
    # Nhãn tương ứng (labels)
    y_train,
    
    # Số lần lặp qua toàn bộ dữ liệu (epoch)
    epochs=5,
    
    # Số mẫu trong mỗi batch
    batch_size=16,
    
    # Tách 20% dữ liệu train để làm validation
    validation_split=0.2,
    
    # verbose=1: hiển thị tiến trình train
    verbose=1
)

Epoch 1/5
10/10 ━━━━━━━━━━━━━━━━━━━━ 1s 33ms/step - accuracy: 0.4875 - loss: 0.6970 - val_accuracy: 0.4750 - val_loss: 0.6919
Epoch 2/5
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 13ms/step - accuracy: 0.5250 - loss: 0.6912 - val_accuracy: 0.4500 - val_loss: 0.6933
Epoch 3/5
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 16ms/step - accuracy: 0.5437 - loss: 0.6872 - val_accuracy: 0.4250 - val_loss: 0.6940
Epoch 4/5
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 19ms/step - accuracy: 0.5375 - loss: 0.6843 - val_accuracy: 0.4250 - val_loss: 0.6946
Epoch 5/5
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 15ms/step - accuracy: 0.5312 - loss: 0.6809 - val_accuracy: 0.4250 - val_loss: 0.6947


In [13]:
# Đánh giá mô hình trên tập test
# Trả về giá trị loss và accuracy
test_loss, test_acc = model.evaluate(x_test, y_test, verbose=0)

# In ra loss (làm tròn 4 chữ số thập phân)
print('Test loss =', round(float(test_loss), 4))

# In ra độ chính xác trên tập test
print('Test accuracy =', round(float(test_acc), 4))

# Dự đoán xác suất cho 5 mẫu đầu tiên của tập test
pred = model.predict(x_test[:5])

# In ra kết quả dự đoán
print('\nDự đoán xác suất cho 5 mẫu đầu:')
print(pred)

Test loss = 0.6721
Test accuracy = 0.52
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 98ms/step

Dự đoán xác suất cho 5 mẫu đầu:
[[0.5582968 ]
 [0.5534731 ]
 [0.649818  ]
 [0.57634616]
 [0.5771873 ]]


## Quy tắc phân loại

Vì model dùng **sigmoid**, nên output là **xác suất nằm trong khoảng [0, 1]**.

**Quy tắc phổ biến**
Nếu xác suất ≥ 0.5 → dự đoán lớp 1  
Nếu xác suất < 0.5 → dự đoán lớp 0  

### Ví dụ

0.8 → lớp 1  
0.3 → lớp 0  
0.5 → thường quy về lớp 1  

### Giải thích bản chất

Sigmoid biến output thành xác suất của lớp 1  
P(y = 1) = output  
P(y = 0) = 1 - output  

### Lưu ý quan trọng

Ngưỡng 0.5 không phải lúc nào cũng tối ưu

Trong thực tế có thể điều chỉnh threshold:  
0.7 → giảm false positive  
0.3 → giảm false negative  

“Nếu model đoán 0.49 và 0.51 thì có khác nhau nhiều không?”


## Câu hỏi thảo luận trên lớp

1. Vì sao TensorFlow phù hợp cho Deep Learning hơn NumPy?

2. Tại sao Keras lại giúp việc xây dựng mô hình trở nên dễ dàng hơn?

3. Khi nào nên dùng Google Colab, khi nào nên dùng máy cá nhân có GPU?

4. `tf.Variable` khác gì với tensor hằng?

5. Tại sao `GradientTape` lại quan trọng trong huấn luyện mô hình?

6. Trong Keras, `compile()` và `fit()` đóng vai trò gì?

## Tổng kết

Sau notebook này, người học cần nắm được:

- TensorFlow là hạ tầng tính toán tensor và gradient.
- Keras là API cấp cao để xây dựng và huấn luyện mô hình Deep Learning.
- Colab là lựa chọn rất phù hợp để bắt đầu thực hành.
- `Variable`, tensor operations và `GradientTape` là nền tảng kỹ thuật quan trọng.
- Quy trình chuẩn trong Keras gồm tạo model, `compile()`, `fit()`, `evaluate()` và `predict()`.